### ISSUE TITLE
Lora merge_and_unload doesn't actually change the base_model's weights.

### ISSUE THREAD

**Issue by ElleLeonne** [huggingface/peft#918]
### System Info

Linux
Python 3.11.4
peft 0.5.0 release
transformers 4.33.0

### Who can help?

@pacman100

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [X] My own task or dataset (give details below)

### Reproduction

'''
lora_config = LoraConfig(r=8, lora_alpha=16, lora_dropout=0.1, bias="none", inference_mode=False,
   target_modules=["q_proj", "k_proj", "v_proj", "o_proj", "gate_proj", "down_proj", "up_proj", "wte", "logits"])
----
og_weight = model.base_model.model.layers[0].self_attn.q_proj.weight
model = model.merge_and_unload() # merge lora
new_weight = model.model.layers[0].self_attn.q_proj.weight
assert not torch.allclose(og_weight, new_weight) # Should be different
'''

### Expected behavior

So I'm working on an integration of ReLoRA with Pytorch Lightning. The above script is part of my checkpointing behavior.

I'm training in 8bit, which means I need to save the lora adapter, then reload the model in 16bit (or 32bit), THEN merge and unload the adapter. Followed immediately by saving the model, and reloading it back into 8bit, initializing a brand new LoRA layer, and resuming train.

However, after further testing, it appears to happen even in 16bit mode, which requires zero state hopping.

However, I noticed an issue in that my loss graphs looked identical between epochs, meaning the weights weren't actually changing in the base model. So, I decided to add a check to confirm whether the merge was actually doing anything or not, and it appears as if it's not. Every single permutation of any merge I do results in identical weights in the base model, regardless of how long I train for, or any other factors.

I had considered that maybe my LoRA model wasn't in training mode, but I can indeed confirm that model.training is on during the forward and backward passes. Just incase, I've included my config file, too. I'm hoping that I missed something obvious in one of the update files. I'm knee deep in a dev environment, so I may try to get an example script working to reproduce the issue in a self-contained manner shortly, assuming this isn't something easy that I'm overlooking.

**Comment by ElleLeonne** [comment]
Updating to 0.6.0dev fixed this for me, for both scenarios.
